bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/TreeViewer.java

```diff
@@ -246,6 +247,20 @@ protected void hookControl(Control control) {
 		super.hookControl(control);
 		Tree treeControl = (Tree) control;
 
+		treeControl.addTreeListener(new TreeListener() {
```

Comment:
The listener should be added and removed when setting the variable for the number of levels to expand in order to only have the listener attached when auto-expansion is enabled. In addition, the current implementation would lead to implementations of `AbstractTreeViewer` that do not extend `TreeViewer` to not automatically expand even if the according values have been set. This means, the chage on `AbstractTreeViewer` currently is API breaking, as all extensions of that class need to be adapted to fulfill the changed method contracts.